docs: Build time styles (#22762)
microsoft/fluentui: apps/public-docsite-v9/src/Concepts/BuildTimeStyles.stories.mdx

import { Meta, Source, Link } from '@storybook/addon-docs';
import imageFile from '../../public/style-resolution.png';

<Meta title="Concepts/Developer/Build time styles" />

## Optimize runtime cost of styles

Fluent UI uses the open source [Griffel.js](https://griffel.js.org/) css-in-js styling engine.
Make sure to take a look at the following docs to get started with styling in Fluent UI:

- [Quick start](?path=/story/quick-start)
- [Styling components](?path=/story/concepts-developer-styling-components--page)

While there is nothing wrong with the associated runtime costs of a css-in-js engine, larger and more complex
applications might want to optimize for performance. In Fluent UI, the expensive runtime work only

> happens on the first render of the component.

Reviewer: This jumps from "happens only once" to "Webpack" too quickly.
1. add one more sentence describing something like "this one-time runtime work can be avoided by processing the styles during build time using *build time transforms*"
2. add a link to the *What is being optimized* below?
3. add a new heading "Setup build time transforms"

Author: Good points

      {
        "modules": [{ "moduleSource": "custom-package", "importName": "makeStyles" }]
      }
    ]
  ]
}
```

### What is being optimized

> ⚠️ Style resolution only needs to happen on the initial render of a component. Therefore, without build time
> optimization, the performance after the initial render is comparable. It is perfectly reasonable to follow the
> quickstart guide and introduce build time optimization if/when it is required.

Let's start with a simple example:

```tsx
import { makeStyles } from '@fluentui/react-components';

const useStyles = makeStyles({
  root: { paddingLeft: '1px', display: 'flex' },
});

function Component() {
  const classes = useStyles();

  return <div className={classes.root} />;
}
```

A correct invocation of `makeStyles` creates a styling hook that can be used inside a component. When the `useStyles`
hook is invoked inside a component the styles are resolved and applied to the document. You can look at the image below
which describes what work is done during style resolution. **Note that this work only happens once, during first render**.

<img src={imageFile} alt={'Illustration of work done by style resolution'} />;

> The final result before the CSS rules are inserted into DOM can be done during build time through

Reviewer: nit?
```suggestion
The final result before the CSS rules are inserted into DOM can be prepared during build time through
```

import { Meta, Source, Link } from '@storybook/addon-docs';
import imageFile from '../../public/style-resolution.png';

<Meta title="Concepts/Developer/Build time styles" />

## Optimize runtime cost of styles

> Fluent UI uses the open source [Griffel.js](https://griffel.js.org/) css-in-js styling engine.

Reviewer 2: ```suggestion
Fluent UI uses the open source [Griffel.js](https://griffel.js.org/) CSS-in-JS styling engine.
```

nit

Make sure to take a look at the following docs to get started with styling in Fluent UI:

> - [Quick start](?path=/story/quick-start)

Reviewer 2: ```suggestion
- [Quick start](?path=/docs/concepts-developer-quick-start--page)
```

The link was broken

- [Styling components](?path=/story/concepts-developer-styling-components--page)

> While there is nothing wrong with the associated runtime costs of a css-in-js engine, larger and more complex

Reviewer 2: ```suggestion
While there is nothing wrong with the associated runtime costs of a CSS-in-JS engine, larger and more complex
```

nit

applications might want to optimize for performance. In Fluent UI, the expensive runtime work only
happens on the first render of the component. This one-time work can be further optimized at build time by
pre-computing and transforming styles. You can refer to the [last section](#what-is-being-optimized) of this page for more details
about what the build time optimization does.

> ## Setup build time style transforms

Reviewer 2: I think that it would useful to put the statement from Griffel's README:

- For application developers, please use Webpack loader
- For library developers, please use the Babel preset

This immediately routes to a proper section

### Webpack

```sh
yarn add @griffel/webpack-loader
```

> In your webpack config, simply add the `@griffel/webpack-loader` to the list of rules.

Reviewer 2: ```suggestion
In your Webpack config, simply add the `@griffel/webpack-loader` to the list of rules.
```

nit

```javascript
// webpack.config.js
module: {
  rules: [
    {
      test: /\.(ts|tsx)$/,
      exclude: /node_modules/,
      use: {
        loader: '@griffel/webpack-loader',
      },

>     },

Reviewer 2: I would modify this to show the order of loaders: usually there is already `babel-loader` or `ts-loader`. Applying loaders in the wrong order will cause issues.

  ];
}
```

### Babel

> The babel preset is used internally by the webpack loader. For applications the webpack loader is appropriate

Reviewer 2: ```suggestion
The Babel preset is used internally by the Webpack loader. For applications the Webpack loader is appropriate
```

_nit_

modules that are not transpiled. Any additional language features can also be configured in the same way.

### Babel

The Babel preset is used internally by the Webpack loader. For applications the Webpack loader is appropriate
since it is the most popular bundler.
However, library developers might prefer to use the Babel preset since it is a better fit for tools such as [Rollup](https://rollupjs.org/guide/en/).

```sh
yarn add @griffel/babel-preset
```

Simply create a Babel configuration file such as `.babelrc` and add the preset.

```js
module.exports = {
  presets: [
    [
      '@griffel',
      {
        // see https://github.com/microsoft/griffel/tree/main/packages/webpack-loader#configuring-babel-settings
        babelOptions: {
          presets: ['@babel/preset-typescript'],
        },
      },
    ],
  ],
};
```

#### Typescript (and additional transpilation) support

[Similar to above](#typescript-and-additional-transpilation-support)

#### Importing from a third party package

If Griffel is re-exported from a third party package, it's necessary to configure the preset to use a different

> module source. By default, the preset already handles imports from `@fluentui/react-components`.

Reviewer 3: > By default, the preset already handles imports from `@fluentui/react-components`.

Can you elaborate more on this ? like its hard coded in our babel-preset ?

```js
import { makeStyles } from 'custom-package';
```

```json
{
  "presets": [
    [
      "@griffel/babel",
      {
        "modules": [{ "moduleSource": "custom-package", "importName": "makeStyles" }]
      }
    ]
  ]
}
```

### What is being optimized

> ⚠️ Style resolution only needs to happen on the initial render of a component. Therefore, without build time

> > optimization, the performance after the initial render is comparable. It is perfectly reasonable to follow the

Reviewer 3: > the performance after the initial render is comparable

It took me some time to understand this but it's just me :D. I'd suggest to reword it to something like
"without build time optimization the performance is comparable with 2nd and consecutive render"

> quickstart guide and introduce build time optimization if/when it is required.

Let's start with a simple example:

> ```tsx

Reviewer 3: **suggestion:** 

expand the example description a bit like following:

```tsx
import { makeStyles } from '@fluentui/react-components';

// 1.
const useStyles = makeStyles({
  root: { paddingLeft: '1px', display: 'flex' },
});

function Component() {
   // 2.
  const classes = useStyles();

  return <div className={classes.root} />;
```

What's happening in that example?

1. invocation of `makeStyles` creates a styling hook that will be used inside a component
2. `useStyles` hook call resolves styles which are injected into the document.

You can look at the image below which describes what work is done during style resolution. 

**💡 NOTE:  this work only happens once, during first render**.

import { makeStyles } from '@fluentui/react-components';

const useStyles = makeStyles({
  root: { paddingLeft: '1px', display: 'flex' },
});

function Component() {
  const classes = useStyles();

  return <div className={classes.root} />;
}
```

A correct invocation of `makeStyles` creates a styling hook that can be used inside a component. When the `useStyles`
hook is invoked inside a component the styles are resolved and applied to the document. You can look at the image below
which describes what work is done during style resolution. **Note that this work only happens once, during first render**.

<img src={imageFile} alt={'Illustration of work done by style resolution'} />

> The final result before the CSS rules are inserted into DOM can be prepared during build time through

Reviewer 3: ```suggestion
The final result before the CSS rules are inserted into DOM can be compiled ahead of time during build time through the methods described above.
```

the methods above. Once the styles of our simple example are transformed at build time the resulting bundle contains
a result similar to what is in our diagram. The actual runtime code of `makeStyles` is completely stripped from the

> bundle and replaced with a lightweight function that simply concatenates the CSS classes and inserts them to

Reviewer 3: ```suggestion
bundle and replaced with a lightweight function (`__styles`) that simply concatenates the CSS classes and inserts them to
```

DOM.

> 

Reviewer 3: ```suggestion

**Result of ahead of time compilation:**

```

Make sure to take a look at the following docs to get started with styling in Fluent UI:

- [Quick start](?path=/docs/concepts-developer-quick-start--page)
- [Styling components](?path=/story/concepts-developer-styling-components--page)

While there is nothing wrong with the associated runtime costs of a CSS-in-JS engine, larger and more complex
applications might want to optimize for performance. In Fluent UI, the expensive runtime work only
happens on the first render of the component. This one-time work can be further optimized at build time by
pre-computing and transforming styles. You can refer to the [last section](#what-is-being-optimized) of this page for more details
about what the build time optimization does.

## Setup build time style transforms

- For application developers please use the [Webpack loader](#webpack)
- For library developers please use the [Babel preset](#babel)

### Webpack

```sh
yarn add @griffel/webpack-loader
```

In your Webpack config, simply add the `@griffel/webpack-loader` to the list of rules.

```javascript
// webpack.config.js
module: {
  rules: [
    {

>       test: /\.(ts|tsx)$/,

Author: yes

      exclude: /node_modules/,
      use: {
        loader: '@griffel/webpack-loader',
      },
    },
    {
      test: /\.(ts|tsx)$/,
      exclude: /node_modules/,
      use: {
        loader: 'babel-loader',
        options: {
          babelOptions: {
            presets: ['@babel/preset-typescript'],
          },
        },

>         // loader: 'ts-loader',

Reviewer 2: This looks a bit confusing. I would rather do:

```js
      use: {
        // or "ts-loader"
        loader: 'babel-loader',
        options: {
```

      },
    },
  ];
}
```

**We recommend adding the Griffel webpack loader after babel-loader or ts-loader.**. The underlying babel transforms
used by the webpack loader are configured by default to use the most basic language features. Therefore to avoid
extra configuration or parsing errors, using the webpack loader after transpilation is the recommended approach.

#### Typescript (and additional transpilation) support

> The Webpack loader [evaluate styles at runtime using Linaria](#build-time-style-evaluation). For typescript

Reviewer 2: ```suggestion
The Webpack loader [evaluate styles at runtime](#build-time-style-evaluation). For typescript
```

Do we want to mention it there?

> it's necessary to install `@babel/preset-typescript` and configure it in the `babelOptions` of the loader. Similarly

Reviewer 2: ```suggestion
it's necessary to install `@babel/preset-typescript` and configure it in the [`babelOptions`](https://github.com/microsoft/griffel/tree/main/packages/webpack-loader#configuring-babel-settings) of the loader. Similarly
```

Is it reasonable to add a link?

any other language features that might be required can also be configured here. While the order of loaders will
ensure that the webpack loader will run on transpiled code (for a specific module), runtime evaluation can potentially require other
modules that are not transpiled. Any additional language features can also be configured in the same way.

### Babel

The Babel preset is used internally by the Webpack loader. For applications the Webpack loader is appropriate
since it is the most popular bundler.
However, library developers might prefer to use the Babel preset since it is a better fit for tools such as [Rollup](https://rollupjs.org/guide/en/).

```sh
yarn add @griffel/babel-preset
```

> Simply create a Babel configuration file such as `.babelrc` and add the preset.

Reviewer 2: AFAIR [`.babelrc.js` is also a thing](https://babeljs.io/docs/en/configuration#javascript-configuration-files), I would rather use it to be able to use comments:

````
```js
module.exports = {
  presets: [
    [
      "@griffel",
      {
        // see https://github.com/microsoft/griffel/tree/main/packages/webpack-loader#configuring-babel-settings
        babelOptions: {
          presets: ["@babel/preset-typescript"]
        }
      }
    ]
  ]
};
```
````

```json
{
  "presets": [
    [
      "@griffel",
      {
        "babelOptions": {
          "presets": ["@babel/preset-typescript"]
        }
      }
    ]
  ]
}
```

> #### Typescript (and additional transpilation) support
> 
> The Babel preset will [evaluate stlyes at runtime using Linaria](#build-time-style-evaluation). For typescript
> it's necessary to install `@babel/preset-typescript` and configure it in the preset. Similarly
> any other language features that might be required can also be configured here. While the order of loaders will
> ensure that the webpack loader will run on transpiled code (for a specific module), runtime evaluation can potentially require other
> modules that are not transpiled. Any additional language features can also be configured in the same way.

Reviewer 2: ```suggestion
```

Is it the same paragraph as above one?

Author: yes, I can remove it.

  return <div className={classes.root} />;
}
```

### Build time style evaluation

Let's consider the following scenario:

```ts
// constants.ts
export const PADDING_TOKEN = '1px';

// common.ts
export const commonStyles = () => ({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
});

// styles.ts
import { makeStyles } from '@fluentui/react-components';
import { PADDING_TOKEN } from './constants';
import { commonStyles } from './common';

const useStyles = makeStyles({
  root: { paddingLeft: PADDING_TOKEN, ...commonStyles() },
});
```

It's perfectly fine, and even recommended to reuse common tokens and create style helpers across an application. It's
one of the main benefits of using CSS-in-JS. However, this means that the build time transforms which are described
above are not trivial to compute because code needs to be evaluated to know what styles to transform. In the example
above, in order to transform the `styles.ts` file, the code needs to be executed/evaluated by importing the extra
modules that it depends on (`constants.ts` and `common.ts`).

Griffel uses
[style evaluation from Linaria](https://github.com/callstack/linaria).

> The build-time evaluation happens as a part of the Griffel Babel preset. All styles that require evaluation will be

Reviewer 2: ```suggestion
The build-time evaluation happens as a part of the Babel transforms in Griffel. All styles that require evaluation will be
```

not sure if mentioning the preset makes it better
